src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,807 @@
+
+            return MsQuicStatusCodes.Success;
+        }
+
+        private uint HandleEventPeerSendShutdown()
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            bool shouldComplete = false;
+
+            lock (_sync)
+            {
+                // This event won't occur within the middle of a receive.
+                if (NetEventSource.IsEnabled) NetEventSource.Info("Completing resettable event source.");
+
+                if (_readState == ReadState.None)
+                {
+                    shouldComplete = true;
+                }
+
+                _readState = ReadState.ReadsCompleted;
+            }
+
+            if (shouldComplete)
+            {
+                _receiveResettableCompletionSource.Complete(0);
+            }
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+            return MsQuicStatusCodes.Success;
+        }
+
+        private uint HandleEventSendComplete(ref StreamEvent evt)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            CleanupSendState();
+
+            // TODO throw if a write was canceled.
+            uint errorCode = evt.Data.SendComplete.Canceled;
+
+            bool shouldComplete = false;
+            lock (_sync)
+            {
+                if (_sendState == SendState.None)
+                {
+                    _sendState = SendState.Finished;
+                    shouldComplete = true;
+                }
+            }
+
+            if (shouldComplete)
+            {
+                _sendResettableCompletionSource.Complete(MsQuicStatusCodes.Success);
+            }
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+            return MsQuicStatusCodes.Success;
+        }
+
+        private void CleanupSendState()
+        {
+            _sendHandle.Free();
+            _bufferArrays[0].Dispose();
+        }
+
+        private void SetCallbackHandler()
+        {
+            _handle = GCHandle.Alloc(this);
+
+            _callback = new StreamCallbackDelegate(NativeCallbackHandler);
+            MsQuicApi.Api._setCallbackHandlerDelegate(
+                _ptr,
+                _callback,
+                GCHandle.ToIntPtr(_handle));
+        }
+
+        // TODO prevent overlapping sends or consider supporting it.
+        private unsafe ValueTask SendAsync(
+           ReadOnlyMemory<byte> buffer,
+           QUIC_SEND_FLAG flags)
+        {
+            if (buffer.IsEmpty)
+            {
+                if ((flags & QUIC_SEND_FLAG.FIN) == QUIC_SEND_FLAG.FIN)
+                {
+                    // Start graceful shutdown sequence if passed in the fin flag and there is an empty buffer.
+                    MsQuicApi.Api._streamShutdownDelegate(_ptr, (uint)QUIC_STREAM_SHUTDOWN_FLAG.GRACEFUL, errorCode: 0);
+                }
+                return default;
+            }
+
+            MemoryHandle handle = buffer.Pin();
```

Comment:
This is never unpinned on error?

Reply:
I do unpin in some scenarios, but I think I'm missing some. CleanupSendState unpins both the sendHandle and _bufferArrays.